### PR TITLE
Merge SPL

### PR DESCRIPTION
**Pull request by behlendorf**
### Description

This PR merges the zfsonlinux/spl repository in to the zfsonlinux/zfs respository and consists of two parts.

In the first part a merge-spl branch is created in the spl repository.  The purpose of this branch is to slightly reorganize the spl source in order to allow it to merge cleanly with the zfs repository.  Files which are no longer needed were removed and the rest were moved to non-conflicting locations.  This branch was then merged in to the zfs repository to preserve the entire commit history.

The second portion of this PR makes the required changes to integrate the relocated spl kernel module in to the zfs repositories build system.  After this point the master branch and 0.8.x and newer releases will not depend on an external spl repository.

NOTE: This PR has been squashed for the purposes of automated testing.  The [merge-spl](https://github.com/behlendorf/zfs/tree/merge-spl) branch which preserves all the history, but is otherwise identical, will be used for the real merge.

### Motivation and Context

Having the spl source code split for the zfs source code is an unnecessary complication for developers, packagers, and end users.  Bringing them together eliminates an entire class of issue which can result when using mismatched versions.

### How Has This Been Tested?

Locally built and tested the following:
* in-tree build and run of the ZTS
* kmod-tracking RPM pacakges
* kmods2 RPM packages

Pending results of the buildbot to indentify any missing build system issues.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by fling-** [comment]
Are the licensing issues expected?

File: module/icp/algs/edonr/edonr.c [hunk 1]
```diff
@@ -30,11 +30,7 @@
  */
 
 /* determine where we can get bcopy/bzero declarations */
-#ifdef	_KERNEL
-#include <sys/systm.h>
-#else
-#include <strings.h>
-#endif
+#include <sys/strings.h>
```
Comment (ahrens) [inline module/icp/algs/edonr/edonr.c]: looks like we aren't really "determining" anything anymore.

Comment (behlendorf) [reply module/icp/algs/edonr/edonr.c]: Indeed.  I'll drop the comment when I'm here.

**Comment by tonyhutter** [comment]
@behlendorf are there any files in particular you want us to review?

**Comment by behlendorf** [comment]
@tonyhutter I belive the most helpful thing would be if you could test the PR locally in your environment and make sure everything works as expected.  The buildbot does provide a substanable amount of build/test coverage but there's still value in manually kicking the tires so to speak.

**Comment by tonyhutter** [comment]
Kmem tracking still works (mem leaks added for testing):
```
[ 2792.253044] ZFS: Unloaded module v0.7.0-1401_g63fd8a3dc (DEBUG mode)
[ 2792.536228] kmem leaked 3300/8944212 bytes
[ 2792.536231] address          size  data             func:line
[ 2792.536248] 000000007505a1ef 100   03440b9200000000 zfs_ioc_pool_stats:1682
[ 2792.536252] 000000003c95065e 200   ee3b0dab00000000 zfs_ioc_pool_stats:1683
[ 2792.536254] 000000005559e127 100   83440b92736f2030 zfs_ioc_pool_stats:1682
[ 2792.536257] 00000000643ee585 200   ee3b0dabb0976aff zfs_ioc_pool_stats:1683
```

**Comment by drescherjm** [comment]
This seems to have broken the gentoo 9999 ebuilds. I assume spl-9999 has to go away. 

* Preparing to run callback: "emerge --oneshot @module-rebuild sys-fs/zfs".....


 * IMPORTANT: 2 config files in '/etc/portage' need updating.
Calculating dependencies
 * IMPORTANT: 1 news items need reading for repository 'gentoo'.
 * Use eselect news read to view new items.

 * See the CONFIGURATION FILES and CONFIGURATION FILES UPDATE TOOLS
 * sections of the emerge man page to learn how to update config files.
... done!
>>> Verifying ebuild manifests
>>> Jobs: 0 of 3 complete, 1 running                Load avg: 4.59, 3.79, 2.51
>>> Emerging (1 of 3) sys-kernel/spl-9999::gentoo
>>> Jobs: 0 of 3 complete, 1 running                Load avg: 4.59, 3.79, 2.51
>>> Failed to emerge sys-kernel/spl-9999, Log file:
>>> Jobs: 0 of 3 complete, 1 running                Load avg: 4.30, 3.74, 2.50
>>>  '/var/tmp/portage/sys-kernel/spl-9999/temp/build.log'
>>> Jobs: 0 of 3 complete, 1 running                Load avg: 4.30, 3.74, 2.50
>>> Jobs: 0 of 3 complete, 1 running, 1 failed      Load avg: 4.30, 3.74, 2.50
>>> Jobs: 0 of 3 complete, 1 failed                 Load avg: 4.30, 3.74, 2.50

Calculating dependencies  *** Resuming merge...
.... done!
>>> Jobs: 0 of 2 complete, 1 running                Load avg: 4.30, 3.74, 2.50
>>> Emerging (1 of 2) sys-fs/zfs-kmod-9999::gentoo
>>> Jobs: 0 of 2 complete, 1 running                Load avg: 4.30, 3.74, 2.50

**Comment by behlendorf** [comment]
@drescherjm that's going to be a question for how @ryao want's to handle this.

**Comment by drescherjm** [comment]
I expected that..

**Comment by ryao** [comment]
@drescherjm @behlendorf It is fixed. I am leaving sys-kernel/spl-9999 around for the time being. It is still useful if anyone wants to use it to build an old version by specifying the commit, but I plan to remove it when 0.8.0 is released.

**Comment by drescherjm** [comment]
Thanks. I am testing now.

**Pull request closed without merging**